### PR TITLE
[loader] Call managed resolving events for the default ALC

### LINKED ISSUE #55921: [mono] Crash in roslyn due to null AssemblyLoadContext.GetLoadContext result
### Description

After upgrading my Mono-based runtime for s390x to preview6, using analyzers with roslyn now crashes, while it used to work in preview5. 

### Configuration

.NET 6 preview6 re-built to target s390x.

### Regression?

Yes, this did work with preview5.   Reverting the patch from https://github.com/dotnet/runtime/pull/53308 fixes the problem.

### Other information

The symptom is error messages along those lines:

```
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.codeanalysis.csharp.codestyle/3.10.0/analyzers/dotnet/cs/Microsoft.CodeAnalysis.CSharp.CodeStyle.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.codeanalysis.csharp.codestyle/3.10.0/analyzers/dotnet/cs/Microsoft.CodeAnalysis.CodeStyle.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.codeanalysis.netanalyzers/6.0.0-rc1.21356.1/analyzers/dotnet/cs/Microsoft.CodeAnalysis.CSharp.NetAnalyzers.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.codeanalysis.netanalyzers/6.0.0-rc1.21356.1/analyzers/dotnet/cs/Microsoft.CodeAnalysis.NetAnalyzers.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.dotnet.codeanalysis/6.0.0-beta.21359.3/analyzers/Microsoft.DotNet.CodeAnalysis.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/microsoft.net.illink.analyzers/6.0.100-preview.6.21362.3/analyzers/dotnet/cs/ILLink.RoslynAnalyzer.dll : Object reference not set to an instance of an object
error CS8034: Unable to load Analyzer assembly /home/uweigand/.nuget/packages/stylecop.analyzers.unstable/1.2.0.304/analyzers/dotnet/cs/StyleCop.Analyzers.dll : Object reference not set to an instance of an object
```

triggered in roslyn's `Microsoft.CodeAnalysis.DefaultAnalyzerAssemblyLoader:LoadFromPathImpl` because `AssemblyLoadContext.GetLoadContext` now returns null for the default ALC.

This is related to https://github.com/dotnet/runtime/pull/53308.  Before that patch, the default ALC used to be initialized in `System.Runtime.Loader.AssemblyLoadContext:InternalInitializeNativeALC` when loading the `libSystem.Native` library on startup.

This routine was called out of the default `AssemblyLoadContext` constructor (which actually allocates the handle) invoked when `GetAssemblyLoadContext` is called with a `IntPtr.Zero` handler, which happens in the `MonoResolveUnmanagedDll` handler, which in turn was called from `netcore_resolve_with_load`.

However, the above PR added this early exit to `netcore_resolve_with_load`:

```
        if (!mono_gchandle_get_target_internal (alc->gchandle))
                return NULL;
```
and now `MonoResolveUnmanagedDll` is never called and therefore we never actually seem to get a non-null handle.

Not sure what the proper fix is, but reverting the PR makes the problem go away.

@CoffeeFlux, @vargaz, @lambdageek can you have a look?

### PR DESCRIPTION
**Pull request by lambdageek**
Refine the work in https://github.com/dotnet/runtime/commit/f70b5b7aef21a70478e9b86dba7d996a57a9add9

The issue is that even if we know that the ALC gchandle points to null in native, we can't automatically skip the call to the managed resolving event (for native library, for example) because the native event is indirectly responsible for creating the default ALC managed object.

Instead, check in native if the gchandle is equal to the default ALC's gchandle (which is initially allocated with a null target) and if so pass `IntPtr.Zero` to the managed code, which will call `AssemblyLoadContext.GetAssemblyLoadContext (IntPtr gch)` which in turn will construct the managed object for the default ALC.

Fixes https://github.com/dotnet/runtime/issues/55921

### CODE REVIEW

**Comment by lambdageek** [comment]
@uweigand can you try this on the s390 port and check whether it resolves the issue?  Also do you know what might be a good way to do a standalone repro?

File: src/mono/mono/metadata/native-library.c [hunk 1]
```diff
@@ -708,7 +712,14 @@ netcore_resolve_with_resolving_event (MonoAssemblyLoadContext *alc, MonoAssembly
 	goto_if_nok (error, leave);
 
 	gpointer gchandle;
```
Comment (uweigand) [inline src/mono/mono/metadata/native-library.c:714]: Don't you also have to remove the 
```
   if (!mono_gchandle_get_target_internal (alc->gchandle))
		return NULL;
```
in this function (line 698 above)?

Comment (lambdageek) [reply src/mono/mono/metadata/native-library.c:714]: yes! thanks!

**Comment by uweigand** [comment]
> @uweigand can you try this on the s390 port and check whether it resolves the issue?

Sure, I'll try this (latest version) and let you know ...

>Also do you know what might be a good way to do a standalone repro?

Unfortunately not really, I was only seeing this with roslyn analzers.  Not sure what exactly those do to trigger the issue ...

**Comment by uweigand** [comment]
@lambdageek This patch does indeed fix the problem I was seeing, thanks!

File: src/mono/mono/metadata/assembly-load-context.c [hunk 2]
```diff
@@ -457,7 +454,14 @@ invoke_resolve_method (MonoMethod *resolve_method, MonoAssemblyLoadContext *alc,
 
 	MonoReflectionAssemblyHandle assm;
 	gpointer gchandle;
-	gchandle = (gpointer)alc->gchandle;
+	/* for the default ALC, pass NULL to ask for the Default ALC - see
```
Comment (vargaz) [inline src/mono/mono/metadata/assembly-load-context.c:457]: This could use a helper function.

**Review by vargaz — APPROVED** [review]

**Review by naricc — APPROVED** [review]

**Pull request merged**